### PR TITLE
[SYCL] Use `std::array` as storage for `sycl::vec` on device

### PR DESCRIPTION
**Pull request by uditagarwal97**
Replaces https://github.com/intel/llvm/pull/13270

Changing the storage to std::array instead of Clang's extension fixes strict ansi-aliasing violation and simplifies device code.

### CODE REVIEW

File: sycl/include/sycl/detail/vector_arith.hpp [hunk 1]
```diff
@@ -228,25 +218,26 @@ class vec_arith : public vec_arith_common<DataT, NumElements> {
...
+    /* we do element-by-element operation on the underlying std::array.  */    \
+    if constexpr (vec_t::IsBfloat16) {                                         \
       for (size_t I = 0; I < NumElements; ++I) {                               \
         /* We cannot use SetValue here as the operator is not a friend of*/    \
         /* Ret on Windows. */                                                  \
-        Ret[I] = static_cast<ocl_t>(                                           \
-            -(vec_data<DataT>::get(detail::VecAccess<vec_t>::getValue(Lhs, I)) \
-                  RELLOGOP vec_data<DataT>::get(                               \
-                      detail::VecAccess<vec_t>::getValue(Rhs, I))));           \
+        Ret[I] = static_cast<ocl_t>(-(Lhs[I] RELLOGOP Rhs[I]));                \
```
Comment (aelovikov-intel) [inline sycl/include/sycl/detail/vector_arith.hpp:227]: Is the comment on line 225 still accurate?

File: sycl/include/sycl/vector_preview.hpp [hunk 2]
```diff
@@ -83,88 +84,39 @@ struct elem {
...
-#endif
-
 template <typename VecT, typename OperationLeftT, typename OperationRightT,
           template <typename> class OperationCurrentT, int... Indexes>
 class SwizzleOp;
 
-template <typename T, int N, typename V = void> struct VecStorage;
+// Element type for relational operator return value.
+template <typename DataT>
+using rel_t = typename std::conditional_t<
```
Comment (aelovikov-intel) [inline sycl/include/sycl/vector_preview.hpp:93]: I think there is a helper for this somewhere.

File: sycl/include/sycl/vector_preview.hpp [hunk 3]
```diff
@@ -778,86 +409,90 @@ class vec : public detail::vec_arith<Type, NumElements> {
...
-    constexpr bool canUseNativeVectorConvert =
+      auto NativeVector = sycl::bit_cast<vector_t>(*this);
+      using ConvertTVecType = typename vec<convertT, NumElements>::vector_t;
+
+      // Whole vector conversion can only be done, if:
+      constexpr bool canUseNativeVectorConvert =
 #ifdef __NVPTX__
-        // - we are not on CUDA, see intel/llvm#11840
-        false &&
+          // - we are not on CUDA, see intel/llvm#11840
```
Comment (aelovikov-intel) [inline sycl/include/sycl/vector_preview.hpp:460]: https://github.com/intel/llvm/issues/11840 is closed. Can we update the comment to something like

```
  TODO: Likely unnecessary as https://github.com/intel/llvm/issues/11840 has been closed already.
```

File: sycl/include/sycl/vector_preview.hpp [hunk 4]
```diff
@@ -778,86 +409,90 @@ class vec : public detail::vec_arith<Type, NumElements> {
...
+      {
+        // Otherwise, we fallback to per-element conversion:
+        for (size_t I = 0; I < NumElements; ++I) {
+          auto val =
+              detail::convertImpl<T, R, roundingMode, 1, OpenCLT, OpenCLR>(
+                  getValue(I));
+          if constexpr (detail::is_byte_v<convertT>)
+            Result.setValue(I, static_cast<convertT>(val));
+          else
+            Result.setValue(I, val);
```
Comment (aelovikov-intel) [inline sycl/include/sycl/vector_preview.hpp:494]: Do we really need `setValue` here? If so, why?

File: sycl/include/sycl/vector_preview.hpp [hunk 5]
```diff
@@ -1005,116 +592,43 @@ class vec : public detail::vec_arith<Type, NumElements> {
...
-  // setValue and getValue.
-  // The "api" functions used by BINOP etc.  These versions just dispatch
-  // using additional int or float arg to disambiguate vec<1> vs. vec<N>
-  // Special proxies as specialization is not allowed in class scope.
   constexpr void setValue(int Index, const DataT &Value) {
-    if (NumElements == 1)
-      setValue(Index, Value, 0);
+#ifdef __SYCL_DEVICE_ONLY__
+    if constexpr (std::is_same_v<DataT, sycl::ext::oneapi::bfloat16>)
+      m_Data[Index] = sycl::bit_cast<DataT>(Value);
```
Comment (aelovikov-intel) [inline sycl/include/sycl/vector_preview.hpp:619]: `Value` is already `DataT`, why do we need to bitcast?

File: sycl/include/sycl/detail/vector_arith.hpp (hunk 1 above)
Comment (uditagarwal97) [reply sycl/include/sycl/detail/vector_arith.hpp:225]: Removed in https://github.com/intel/llvm/pull/14130/commits/ee589312ff65c2b3e95ae6c3e50cf2b5cc0612b5

File: sycl/include/sycl/vector_preview.hpp (hunk 2 above)
Comment (uditagarwal97) [reply sycl/include/sycl/vector_preview.hpp]: Yes, I've replaced `rel_t` with `detail::select_cl_scalar_integral_signed_t`. https://github.com/intel/llvm/pull/14130/commits/ee589312ff65c2b3e95ae6c3e50cf2b5cc0612b5

File: sycl/include/sycl/vector_preview.hpp (hunk 5 above)
Comment (uditagarwal97) [reply sycl/include/sycl/vector_preview.hpp]: True. It's likely a leftover from the previous implementation. I've removed this and inlined `setValue` completely.

File: sycl/include/sycl/vector_preview.hpp (hunk 4 above)
Comment (uditagarwal97) [reply sycl/include/sycl/vector_preview.hpp]: I've inlined `setValue` in https://github.com/intel/llvm/pull/14130/commits/ee589312ff65c2b3e95ae6c3e50cf2b5cc0612b5

**Review by aelovikov-intel — APPROVED** [review]

**Comment by uditagarwal97** [comment]
@intel/llvm-gatekeepers the PR is ready.

**Comment by aelovikov-intel** [comment]
I think this broke CTS in Nightly.

**Comment by uditagarwal97** [comment]
> I think this broke CTS in Nightly.

PR for fix: https://github.com/intel/llvm/pull/14164

**Pull request merged**